### ISSUE TITLE
pyright 1.1.236: Incorrect comparison of overloads involving ParamSpec

### ISSUE THREAD

**Issue by rsokl** [microsoft/pyright#3330]
**Describe the bug**
As of pyright 1.1.236, the following code produces an error under strict mode. Specifically, overload 3 is marked as incompatible with the overloaded implementation, even though they are identical.

The issue seems to be that pyright is comparing `Builds[Type[T2@__call__]] | BuildsWithSig[Type[R@__call__], P@__call__]` against `Builds[(**P@__call__) -> R@__call__] | BuildsWithSig[Any, (*args: Any, **kwargs: Any)]`, whereas it should be comparing the former against itself. 

I've included the full error below.

**To Reproduce**
```python
# pyright: strict

from typing import Any, Callable, Type, TypeVar, Union, overload

from typing_extensions import Literal, ParamSpec, Protocol

T1 = TypeVar("T1", covariant=True)

T2 = TypeVar("T2", bound=Callable[..., Any])
R = TypeVar("R")
P = ParamSpec("P")


class Builds(Protocol[T1]):
    _target_: str


class BuildsWithSig(Builds[T1], Protocol[T1, P]):
    def __init__(self, *args: P.args, **kwds: P.kwargs):
        ...


class PBuilds(Protocol):
    @overload
    def __call__(
        self,
        x: Callable[P, R],
        *,
        sig: Literal[True] = ...,
    ) -> BuildsWithSig[Type[R], P]:
        ...

    @overload
    def __call__(
        self,
        x: T2,
        *,
        sig: Literal[False] = ...,
    ) -> Builds[Type[T2]]:
        ...

    @overload
    def __call__(
        self,
        x: Union[T2, Callable[P, R]],
        *,
        sig: bool,
    ) -> Union[Builds[Type[T2]], BuildsWithSig[Type[R], P]]:
        ...

    def __call__(
        self,
        x: Union[T2, Callable[P, R]],
        *,
        sig: bool = False,
    ) -> Union[Builds[Type[T2]], BuildsWithSig[Type[R], P]]:
        ...
```

The reported error is:

```
error: Overloaded implementation is not consistent with signature of overload 3
    Function return type "Builds[Type[T2@__call__]] | BuildsWithSig[Type[R@__call__], P@__call__]" is incompatible with type "Builds[(**P@__call__) -> R@__call__] | BuildsWithSig[Any, (*args: Any, **kwargs: Any)]"
      Type "Builds[Type[T2@__call__]] | BuildsWithSig[Type[R@__call__], P@__call__]" cannot be assigned to type "Builds[(**P@__call__) -> R@__call__] | BuildsWithSig[Any, (*args: Any, **kwargs: Any)]"
        Type "Builds[Type[T2@__call__]]" cannot be assigned to type "Builds[(**P@__call__) -> R@__call__] | BuildsWithSig[Any, (*args: Any, **kwargs: Any)]"
          TypeVar "T1@Builds" is covariant
            Type "Type[T2@__call__]" cannot be assigned to type "(**P@__call__) -> R@__call__"
          "Builds[Type[T2@__call__]]" is incompatible with protocol "BuildsWithSig[Any, (*args: Any, **kwargs: Any)]"
            TypeVar "P@BuildsWithSig" is invariant (reportGeneralTypeIssues
```

Interestingly, simply flipping the order of the union for argument `x` in overload 3 eliminates the error:

```python
    # ... snippet starting with overload 3

    @overload
    def __call__(
        self,
        x: Union[Callable[P, R], T2],   # <- I flipped this union, which eliminates the error
        *,
        sig: bool,
    ) -> Union[Builds[Type[T2]], BuildsWithSig[Type[R], P]]:
        ...

    def __call__(
        self,
        x: Union[T2, Callable[P, R]],
        *,
        sig: bool = False,
    ) -> Union[Builds[Type[T2]], BuildsWithSig[Type[R], P]]:
        ...
```

**Comment by erictraut** [comment]
The problem is that you've created a generic signature with an ambiguous solution. The type expressions `T2` (which is bound to `Callable[..., Any]`) and `Callable[P, R]` will both match a callable argument. Pyright's TypeVar solver has various heuristics that attempt to pick the best way to disambiguate these situations, but as with all heuristics, they will invariably fail in some cases. The reason this just happened to work previously was due to a bug that was fixed in 1.1.236.

My recommendation is for you to redesign the API to eliminate this ambiguity. Either stick with a traditional TypeVar bound to a Callable or use a ParamSpec. Don't use both.

I'll take a closer look at the heuristics in pyright to see if I can get it to resolve this ambiguity in a way that would help you out in this situation, but these heuristics have already received significant tuning, and it's quite possible that we might just be trading off one failure for another.

**Comment by rsokl** [comment]
> The type expressions T2 (which is bound to Callable[..., Any]) and Callable[P, R] will both match a callable argument [...]  Either stick with a traditional TypeVar bound to a Callable or use a ParamSpec. Don't use both.

Thanks for the explanation and for the recommendation.  I appreciate you looking into the heuristics, and I understand if there is simply not a good trade-off here.

In case it is of interest: I had considered writing `Builds` as  `class Builds(Protocol[T1, P]): ...` so that I would rely only on `ParamSpec` However, one problem with this is that `ParamSpec` cannot be used with overloaded functions. This would lead my users to hit type errors for some very common use cases. `TypeVar` + `ParamSpec` makes it easier to silence these.

```python
def example(func: PBuilds):
    # Using TypeVar + ParamSpec: Builds[Type[dict[Unknown, Unknown]]]
    # Using ParamSpec only: error: ParamSpec cannot be used with overloaded function
    func(dict, sig=False)  # dict and other stdlib functions have overloads
```

I realize that this is hacky and doesn't fully solve the issue, but I am really trying to tune the user experience so that they can benefit from signature completion as much as possible (when they opt-in to it) with as little noise as possible.

**Comment by erictraut** [comment]
I introduced a separate bug fix to the last release (1.1.238) that eliminated the error in your code sample above. 

I've added a new unit test based on your code sample, and I'll try to keep that from regressing in the future. However, as I said, the code is relying on ambiguous behaviors and is therefore fragile in nature. Other type checkers, for example, could easily produce a different result.

**Comment by rsokl** [comment]
>I introduced a separate bug fix to the last release (1.1.238) that eliminated the error in your code sample above.
>I've added a new unit test based on your code sample, and I'll try to keep that from regressing in the future

Thank you for the update on this!

> the code is relying on ambiguous behaviors and is therefore fragile in nature

Understood, I will understand if pyright's behavior changes in the future. If it does, I will simple eliminate this signature-completion behavior from my library.
